Feature: Generate Methods for Resources Having String Interpolation Arguments

krishankumar95 opened the pull request:
**Changes:**
a) Any ResX String containing "{0} {1} ..." interpolation arguments now generate statically typed methods where these args can be passed as parameters. 

b) Any ResX String which does not contain such interpolation args continue to function as is.

**Fixed a bug where:**
a) Identifiers starting with digits are allowed.

Commit:
Added a feature as follows:
a) Any ResX String containing "{0} {1} ..." interploation agruments now generate statically typed methods where these args can be passed as parameters.
b) Any ResX String which does not contain such interpolation args continue to function as is.

Fixed a bug where :
a) Identifers starting with digits are allowed.

Commit:
Added unit and integration tests for Method Generation Feature.

krishankumar95 commented:
Usage: 
Consider a resx file with data elements snippet as follows:

```
  <data name=""Method1WithTwoParams"" xml:space=""preserve"">
    <value>Parameter1:{0},Parameter2:{1}</value>
  </data>
  <data name=""Method2WithThreeParams"" xml:space=""preserve"">
    <value>Parameter1:{0},Parameter2:{1},Parameter3{2}</value>
  </data>
  <data name=""JustAProperty"" xml:space=""preserve"">
    <value>Plain property with no params</value>
  </data>
  ```
  
  Generates the following code: 
  ```
    /// <summary>
    /// Looks up a localized string similar to Parameter1:{0},Parameter2:{1}.
    /// </summary>
    public static string? Method1WithTwoParams(string param0, string param1)=> String.Format(CultureInfo, ResourceManager.GetString(""Method1WithTwoParams"", CultureInfo), param0, param1);

    /// <summary>
    /// Looks up a localized string similar to Parameter1:{0},Parameter2:{1},Parameter3{2}.
    /// </summary>
    public static string? Method2WithThreeParams(string param0, string param1, string param2)=> String.Format(CultureInfo, ResourceManager.GetString(""Method2WithThreeParams"", CultureInfo), param0, param1, param2);

    /// <summary>
    /// Looks up a localized string similar to Plain property with no params.
    /// </summary>
    public static string? JustAProperty => ResourceManager.GetString(nameof(JustAProperty), CultureInfo);
 ```
    
Which can be used as following:
```
// NET 6 Console Sample
using GeneratedNamespaceName;

Console.WriteLine(GeneratedClassName.Method1WithTwoParams("hello","world"));
Console.WriteLine(GeneratedClassName.Method2WithThreeParams("hello","world","!!!"));
Console.WriteLine(GeneratedClassName.JustAProperty);

```

ycanardeau commented on VocaDb.ResXFileCodeGenerator.Tests/GithubIssues/FeatureRequests/MethodGeneratorTests.cs:
```diff
@@ -0,0 +1,144 @@
+	[Fact]
+	public void Generate_StringBuilder_Name_NotValidIdentifier()
+	{
+		var text = @"<?xml version=""1.0"" encoding=""utf-8""?>
```
Can you please use [raw string literals](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/tokens/raw-string)?

ycanardeau commented on VocaDb.ResXFileCodeGenerator/StringBuilderGenerator.cs:
```diff
@@ -224,14 +237,44 @@ private static IEnumerable<(string key, string value, IXmlLineInfo line)>? ReadR
+
+		if (numParams == 0)
+		{
+			builder.Append(indent);
+			builder.Append("public ");
+			builder.Append(options.StaticMembers ? "static " : string.Empty);
+			builder.Append("string");
+			builder.Append(options.NullForgivingOperators ? null : "?");
+			builder.Append(" ");
+			builder.Append(memberName);
+			return true;
+		}
+		else
+		{
+			builder.Append(indent);
+			builder.Append("public ");
+			builder.Append(options.StaticMembers ? "static " : string.Empty);
+			builder.Append("string");
+			builder.Append(options.NullForgivingOperators ? null : "?");
+			builder.Append(" ");
+			builder.Append(memberName);
```
Is it possible to move these lines out of the if and else blocks, or extract to a method?

ycanardeau commented on VocaDb.ResXFileCodeGenerator.Tests/GithubIssues/FeatureRequests/MethodGeneratorTests.cs:
```diff
@@ -0,0 +1,144 @@
+    /// <summary>
+    /// Looks up a localized string similar to Parameter1:{0},Parameter2:{1}.
+    /// </summary>
+    public static string? Method1WithTwoParams(string param0, string param1)=> String.Format(CultureInfo, ResourceManager.GetString(""Method1WithTwoParams"", CultureInfo), param0, param1);
```
It's trivial, but can you please add a space before `=>` in generated code?

ycanardeau commented on VocaDb.ResXFileCodeGenerator/VocaDb.ResXFileCodeGenerator.csproj:
```diff
@@ -15,7 +15,7 @@
     <NoWarn>$(NoWarn);NU5128</NoWarn>
     <DevelopmentDependency>true</DevelopmentDependency>
-    <Version>3.2.0</Version>
+    <Version>3.3.0-alpha</Version>
```
I forgot to push the latest commit, so there might be a merge conflict here.

ycanardeau commented on VocaDb.ResXFileCodeGenerator.Tests/GithubIssues/Issue3/GeneratorTests.cs:
```diff
@@ -70,8 +70,8 @@ public void Generate_StringBuilder_Name_NotValidIdentifier()
   </resheader>
-  <data name=""Invalid identifier {0}"" xml:space=""preserve"">
-    <value>String '{0}' is not a valid identifier.</value>
+  <data name=""1InvalidId"" xml:space=""preserve"">
+    <value>Identifier cannot start with a number.</value>
```
Is there any particular reason why you changed these unit tests rather than adding new ones?

ycanardeau reviewed (commented):
> a) Any ResX String containing "{0} {1} ..." interpolation arguments now generate statically typed methods where these args can be passed as parameters.

Is there any way to escape "{0} {1} ..."? Since this seems to be a breaking change, wouldn't it be better to make this an opt-in feature, just in case? Perhaps it would be better to prefix these methods with `Get` or `Format` to distinguish them from properties (or better to generate both properties and methods).

Commit:
Updates:
a) Param types to obj from str
b) Fixed Bug where repeated use of same args in string generates excess amount of params.

krishankumar95 commented:
Closing current PR; Will reopen a new PR with resolved comments.
